sdk/servicebus/Azure.Messaging.ServiceBus/src/Primitives/ServiceBusRetryPolicy.cs

```diff
@@ -185,49 +186,44 @@ public abstract class ServiceBusRetryPolicy
             throw new TaskCanceledException();
         }
 
-        internal void SetServerBusy(string exceptionMessage)
+        private void SetServerBusy(string exceptionMessage, CancellationToken cancellationToken)
         {
             // multiple call to this method will not prolong the timer.
-            if (encounteredServerBusy)
+            if (Interlocked.CompareExchange(ref serverBusyLock, 1, 0) != 0)
             {
                 return;
             }
 
-            lock (serverBusyLock)
-            {
-                if (!encounteredServerBusy)
-                {
-                    encounteredServerBusy = true;
-                    ServerBusyExceptionMessage = string.IsNullOrWhiteSpace(exceptionMessage) ?
-                        Resources.DefaultServerBusyException : exceptionMessage;
-                    IsServerBusy = true;
-                    _ = ScheduleResetServerBusy();
-                }
-            }
+            ServerBusyExceptionMessage = string.IsNullOrWhiteSpace(exceptionMessage) ?
+                Resources.DefaultServerBusyException : exceptionMessage;
+            IsServerBusy = true;
+            Volatile.Write(ref serverBusyLock, 0);  // release
+            _ = ScheduleResetServerBusy(cancellationToken);
         }
 
-        internal void ResetServerBusy()
+        private void ResetServerBusy()
         {
-            if (!encounteredServerBusy)
+            if (Interlocked.CompareExchange(ref serverBusyLock, 1, 0) != 0)
```

Comment:
I wonder if `Volatile.Read` would suffice here, as we don't want to actually set the value and it still ensures a memory barrier.   Functionally, I believe they're equivalent, but the `Read` more clearly signals the intent.

Reply:
The idea was here to set the value to make sure we are not setting it to non-busy if simultaneously another thread sets it to busy.